NitroxClient/GameLogic/Spawning/Bases/GhostEntitySpawner.cs

```diff
@@ -0,0 +1,155 @@
+        }
+
+        return ghost;
+    }
+
+    public static IEnumerator RestoreGhost(Transform parent, GhostEntity ghostEntity, TaskResult<Optional<GameObject>> result = null)
+    {
+        Log.Debug($"Restoring ghost {ghostEntity}");
+
+        if (!DefaultWorldEntitySpawner.TryGetCachedPrefab(out GameObject prefab, classId: ghostEntity.ClassId))
+        {
+            TaskResult<GameObject> prefabResult = new();
+            yield return DefaultWorldEntitySpawner.RequestPrefab(ghostEntity.ClassId, prefabResult);
+            if (!prefabResult.Get())
+            {
+                Log.Debug($"Couldn't find a prefab for ghost of ClassId {ghostEntity.ClassId}");
+                yield break;
+            }
+            prefab = prefabResult.Get();
+        }
+
+        bool isInBase = parent.TryGetComponent(out Base @base);
+
+        GameObject ghostObject = UnityEngine.Object.Instantiate(prefab);
+        Transform ghostTransform = ghostObject.transform;
+        MoveGhostToTransform(ghostEntity, ghostTransform);
+
+        ConstructableBase constructableBase = ghostObject.GetComponent<ConstructableBase>();
+        GameObject ghostModel = constructableBase.model;
+        BaseGhost baseGhost = ghostModel.GetComponent<BaseGhost>();
+        Base ghostBase = ghostModel.GetComponent<Base>();
+        bool isBaseDeconstructable = ghostObject.name.Equals("BaseDeconstructable(Clone)");
+        if (isBaseDeconstructable && ghostEntity.TechType != null)
+        {
+            constructableBase.techType = ghostEntity.TechType.ToUnity();
+        }
+        constructableBase.constructedAmount = ghostEntity.ConstructedAmount;
+
+        baseGhost.SetupGhost();
+
+        yield return GhostMetadataApplier.ApplyMetadataToGhost(baseGhost, ghostEntity.Metadata, @base);
+
+        // Necessary to wait for BaseGhost.Start()
+        yield return null;
+        // Verify that the metadata didn't destroy the GameObject
+        if (!ghostObject)
+        {
+            yield break;
+        }
+
+        BuildEntitySpawner.ApplyBaseData(ghostEntity.BaseData, ghostBase);
+        ghostBase.OnProtoDeserialize(null);
+        if (ghostBase.cellObjects != null)
+        {
+            Array.Clear(ghostBase.cellObjects, 0, ghostBase.cellObjects.Length);
+        }
+        ghostBase.FinishDeserialization();
+
+        if (isInBase)
+        {
+            @base.SetPlacementGhost(baseGhost);
+            baseGhost.targetBase = @base;
+            @base.RegisterBaseGhost(baseGhost);
+        }
+        else
+        {
+            ghostTransform.parent = parent;
+        }
+        constructableBase.SetGhostVisible(false);
+
+        Log.Debug(BuildEntitySpawner.GetBaseData(ghostBase));
+        if (!isBaseDeconstructable)
+        {
+            baseGhost.Place();
+        }
+
+        if (isInBase)
+        {
+            ghostTransform.parent = parent;
+            MoveGhostToTransform(ghostEntity, ghostTransform);
```

Comment:
I thought I had but I apparently didn't write a comment to explain how important the exact instruction ordering is. Yes it is because of baseGhost.Place() instruction in the upper if (which is completely independent from the other ifs)